core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,616 @@
+    },
+};
+
+mod packet_container;
+
+/// Value chosen because it was used historically, at some point
+/// was found to be optimal. If we need to improve performance
+/// this should be evaluated with new stage.
+const FORWARD_BATCH_SIZE: usize = 128;
+
+/// Addresses to use for forwarding.
+#[derive(Clone)]
+pub struct ForwardingAddresses {
+    /// Forwarding address for TPU (non-votes)
+    pub tpu: Option<SocketAddr>,
+    /// Forwarding address for votes
+    pub tpu_vote: Option<SocketAddr>,
+}
+
+pub trait ForwardAddressGetter: Send + Sync + 'static {
+    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses;
+}
+
+impl<T: LikeClusterInfo> ForwardAddressGetter for (T, Arc<RwLock<PohRecorder>>) {
+    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses {
+        ForwardingAddresses {
+            tpu: next_leader(&self.0, &self.1, |node| node.tpu_forwards(protocol)).map(|(_, s)| s),
+            tpu_vote: next_leader_tpu_vote(&self.0, &self.1).map(|(_, s)| s),
+        }
+    }
+}
+
+struct VoteClient {
+    udp_socket: UdpSocket,
+}
+
+impl VoteClient {
+    fn new() -> Self {
+        Self {
+            udp_socket: bind_to_unspecified().unwrap(),
+        }
+    }
+
+    fn send_batch(&self, input_batch: &mut Vec<(Vec<u8>, SocketAddr)>) {
+        let mut batch = Vec::with_capacity(FORWARD_BATCH_SIZE);
+        core::mem::swap(&mut batch, input_batch); // why do we swap?
+        let _res = batch_send(&self.udp_socket, &batch);
+    }
+}
+
+/// Forwards packets to current/next leader.
+pub struct ForwardingStage<F: ForwardAddressGetter> {
+    receiver: Receiver<(BankingPacketBatch, bool)>,
+    packet_container: PacketContainer,
+
+    root_bank_cache: RootBankCache,
+    forward_address_getter: F,
+    connection_cache: Arc<ConnectionCache>,
+    vote_client: VoteClient,
+    data_budget: DataBudget,
+
+    metrics: ForwardingStageMetrics,
+}
+
+impl<F: ForwardAddressGetter> ForwardingStage<F> {
+    pub fn spawn(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        connection_cache: Arc<ConnectionCache>,
+        root_bank_cache: RootBankCache,
+        forward_address_getter: F,
+        data_budget: DataBudget,
+    ) -> JoinHandle<()> {
+        let forwarding_stage = Self::new(
+            receiver,
+            connection_cache,
+            root_bank_cache,
+            forward_address_getter,
+            data_budget,
+        );
+        Builder::new()
+            .name("solFwdStage".to_string())
+            .spawn(move || forwarding_stage.run())
+            .unwrap()
+    }
+
+    fn new(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        connection_cache: Arc<ConnectionCache>,
+        root_bank_cache: RootBankCache,
+        forward_address_getter: F,
+        data_budget: DataBudget,
+    ) -> Self {
+        Self {
+            receiver,
+            packet_container: PacketContainer::with_capacity(4 * 4096),
```

Comment:
Could you document the choice of this number? Or it has been taken from Forwarder and it is currently unclear how was chosen?

Reply:
just large enough that we can do some reasonable priority-ordering, but not so large that we use a lot of memory - no real justification here.